far/common/2d/segment.hpp

```diff
@@ -0,0 +1,93 @@
+﻿#ifndef SEGMENT_HPP_DC38F7F7_0E1B_486E_8368_2ACE6415D234
+#define SEGMENT_HPP_DC38F7F7_0E1B_486E_8368_2ACE6415D234
+#pragma once
+
+/*
+segment.hpp
+
+*/
+/*
+Copyright © 2018 Far Group
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+1. Redistributions of source code must retain the above copyright
+   notice, this list of conditions and the following disclaimer.
+2. Redistributions in binary form must reproduce the above copyright
+   notice, this list of conditions and the following disclaimer in the
+   documentation and/or other materials provided with the distribution.
+3. The name of the authors may not be used to endorse or promote products
+   derived from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+//----------------------------------------------------------------------------
+
+template<typename T>
+struct segment_length_t
```

Comment:
Supposedly this is to resolve ambiguous ctors (Begin, End) vs (Begin, Length)?
Do we actually need both?
Looks like (Begin, Length) is used only in one place now.

Reply:
Correct. So far, yes, it's used only in the `Editor`. But it makes a good concept (pun unintended). A `segment` can be considered as either (start, end) or (start, length), and one should never be limited to one particular representation, neither in accessors, nor in constructors. If you think it proper, I can add `begin()` and `end()` methods to make it fully symmetrical.

Let's keep it.